src/ambianic/config_mgm.py

```diff
@@ -0,0 +1,163 @@
+"""Centralized configuration manager."""
+
+import os
+import logging
+import threading
+import yaml
+from inotify_simple import INotify, flags
+
+
+class ConfigurationManager:
+    """Configuration manager handles configuration centrally and
+       notify via callbacks of changes
+    """
+
+    def __init__(self, work_dir=None):
+
+        self.CONFIG_FILE = "config.yaml"
+        self.SECRETS_FILE = "secrets.yaml"
+
+        self.lock = threading.RLock()
+        self.__config = None
+        self.watch_thread = None
+        self.watch_event = threading.Event()
+        self.log = logging.getLogger()
+        self.handlers = []
+
+        if work_dir is not None:
+            self.load(work_dir)
+
+    def stop(self):
+        """Stop the config manager"""
+        self.handlers = []
+        with self.lock:
+            self.__config = None
+        self.watch_stop()
+        if self.watch_thread is not None:
+            self.watch_thread.join()
+            self.watch_thread = None
+
+    def register_handler(self, callback):
+        """Register a callback to trigger when there is a configuration update"""
+        self.handlers.append(callback)
+
+    def unregister_handler(self, callback):
+        """Remove a callback from the configuration updates handlers"""
+        self.handlers.remove(callback)
+
+    def __watcher(self):
+        """Watch for file changes"""
+        inotify = INotify()
+        wd = inotify.add_watch(self.work_dir, flags.MODIFY)
+        while not self.watch_event.is_set():
+            for event in inotify.read(timeout=100, read_delay=100):
+                for filename in [self.CONFIG_FILE, self.SECRETS_FILE]:
+                    if event.name == filename:
+                        self.log.info("File change detected: %s", filename)
+                        self.load(self.work_dir)
+                        break
+        # stop watching
+        inotify.rm_watch(wd)
+
+    def watch_start(self):
+        """Start watching fs for changes"""
+        if self.watch_thread is None:
+            self.watch_event.clear()
+            self.watch_thread = threading.Thread(target=self.__watcher)
+            self.watch_thread.start()
+
+    def watch_stop(self):
+        """Stop watching fs for changes"""
+        self.watch_event.set()
+
+    def save(self, config):
+        """Save configuration to file"""
+        if config is not None:
+            return
+
+        with open(self.get_config_file(), 'w') as fh:
+            yaml.dump(config, fh, default_flow_style=False)
```

Comment:
**Optional topic** related to the question I asked in a different comment. Can we try [Ruyaml](https://pypi.org/project/ruamel.yaml/) or another a drop-in replacement for PyYaml that preserves comments and structure of yaml file after load and save.
I think we will need a more granular API for config changes instead of a single YAML file dump over HTTP.
I am not sure if it will make sense to send back and forth the full yaml file between UI and edge. 
Full YAML file upload might make sense for fleet management for initial device setup, but after that, granular APIs would be probably more practical. I guess we will find out with time.

Reply:
Totally agree, see also my comment above. I will go through the sections of the config and identify appropriate APIs to expose with validation.